File: Java/java_labwork/12-11-25/Shape.java
package lab;

class Shape
{
	String color;
	double area;

	Shape() // default constructor
	{
		this.color = "Red";
		this.area = 20;
	}

	Shape(String color, double area) // parameterized constructor
	{
		this.color = color;
		this.area = area;
	}

	// Getters and setters
	public String getColor() {
		return color;
	}
	public void setColor(String color) {
		this.color = color;
	}
	public double getArea() {
		return area;
	}
	public void setArea(double area) {
		this.area = area;
	}
	
	double calArea()
	{
		return area;
	}

	void display() {
		System.out.println("Color: " + this.color);
		System.out.println("Area: " + this.area);
		System.out.println("Area is: "+this.calArea());
	}
} // Shape ends here


class Circle extends Shape
{
	double radius;

	Circle() // default constructor
	{
		super();
		this.radius = 5.0;
		this.area = Math.PI * radius * radius;
	}

	Circle(String color, double radius) // parameterized constructor
	{
		super(color, Math.PI * radius * radius);
		this.radius = radius;
	}

	public double getRadius() {
		return radius;
	}
	public void setRadius(double radius) {
		this.radius = radius;
		this.area = Math.PI * radius * radius;
	}
	
	double calArea()
	{
		return area;
	}

	void display() {
		super.display();
		System.out.println("Radius: " + this.radius);
		System.out.println();
		System.out.println("Area is: "+this.calArea());
	}
} // Circle ends here


class Rectangle extends Shape
{
	double length;
	double breadth;

	Rectangle() // default constructor
	{
		super();
		this.length = 4;
		this.breadth = 9;
		this.area = length * breadth;
	}

	Rectangle(String color, double length, double breadth) // parameterized constructor
	{
		super(color, length * breadth);
		this.length = length;
		this.breadth = breadth;
	}

	public double getLength() {
		return length;
	}
	public void setLength(double length) {
		this.length = length;
		this.area = this.length * this.breadth;
	}
	public double getBreadth() {
		return breadth;
	}
	public void setBreadth(double breadth) {
		this.breadth = breadth;
		this.area = this.length * this.breadth;
	}
	
	double calArea()
	{
		return area;
	}

	void display() {
		super.display();
		System.out.println("Length: " + this.length);
		System.out.println("Breadth: " + this.breadth);
		System.out.println("Area is: "+this.calArea());
		System.out.println();
	}
} // Rectangle ends here


class Triangle extends Shape
{
	double base;
	double height;

	Triangle() // default constructor
	{
		super();
		this.base = 5.0;
		this.height = 4.0;
		this.area = 0.5 * base * height;
	}

	Triangle(String color, double base, double height) // parameterized constructor
	{
		super(color, 0.5 * base * height);
		this.base = base;
		this.height = height;
	}

	public double getBase() {
		return base;
	}
	public void setBase(double base) {
		this.base = base;
		this.area = 0.5 * base * this.height;
	}
	public double getHeight() {
		return height;
	}
	public void setHeight(double height) {
		this.height = height;
		this.area = 0.5 * this.base * height;
	}
	
	double calArea()
	{
		return area;
	}

	void display() {
		super.display();
		System.out.println("Base: " + this.base);
		System.out.println("Height: " + this.height);
		System.out.println("Area is: "+this.calArea());
		System.out.println();
	}
} // Triangle ends here


class ShapeTest
{
	public static void main(String[] args)
	{
		Shape s1;
		System.out.println("\n--- Shape ---");
		 s1 = new Shape();
		 s1 = new Shape("Yellow", 200.0);
		 s1.display();
		 
		 System.out.println("\n--- Circle  ---");
		 s1 = new Circle("Blue", 10.0);
		 s1.display();
		 
		 System.out.println("\n--- Rectangle  ---");
		 s1 = new Rectangle("Green", 8.0, 6.0);
		 s1.display();
		 
		 System.out.println("\n--- Triangle  ---");
		 s1 = new Triangle("Purple", 5.0, 7.0);
		 s1.display();

		
	} //main ends here
} // ShapeTest ends here

File: c++/Assignment16/Polymorphism by structure/Shape.cpp
#include <iostream>
#include <string>
#include <cmath>
using namespace std;

// -------------------- Shape --------------------
struct Shape
{
protected:
    string color;
    double area;

public:
    // default constructor
    Shape()
    {
        color = "Red";
        area = 20;
    }

    // parameterized constructor
    Shape(string c, double a)
    {
        color = c;
        area = a;
    }

    virtual double calArea()
    {
        return area;
    }

    virtual void display()
    {
        cout << "Color: " << color << endl;
        cout << "Area: " << area << endl;
        cout << "Area is: " << calArea() << endl;
    }
};

// -------------------- Circle --------------------
struct Circle : public Shape
{
protected:
    double radius;

public:
    Circle() : Shape()
    {
        radius = 5.0;
        area = M_PI * radius * radius;
    }

    Circle(string c, double r)
        : Shape(c, M_PI * r * r)
    {
        radius = r;
    }

    double calArea()
    {
        return area;
    }

    void display()
    {
        Shape::display();
        cout << "Radius: " << radius << endl;
        cout << "Area is: " << calArea() << endl << endl;
    }
};

// -------------------- Rectangle --------------------
struct Rectangle : public Shape
{
protected:
    double length;
    double breadth;

public:
    Rectangle() : Shape()
    {
        length = 4;
        breadth = 9;
        area = length * breadth;
    }

    Rectangle(string c, double l, double b)
        : Shape(c, l * b)
    {
        length = l;
        breadth = b;
    }

    double calArea()
    {
        return area;
    }

    void display()
    {
        Shape::display();
        cout << "Length: " << length << endl;
        cout << "Breadth: " << breadth << endl;
        cout << "Area is: " << calArea() << endl << endl;
    }
};

// -------------------- Triangle --------------------
struct Triangle : public Shape
{
protected:
    double base;
    double height;

public:
    Triangle() : Shape()
    {
        base = 5.0;
        height = 4.0;
        area = 0.5 * base * height;
    }

    Triangle(string c, double b, double h)
        : Shape(c, 0.5 * b * h)
    {
        base = b;
        height = h;
    }

    double calArea()
    {
        return area;
    }

    void display()
    {
        Shape::display();
        cout << "Base: " << base << endl;
        cout << "Height: " << height << endl;
        cout << "Area is: " << calArea() << endl << endl;
    }
};

// -------------------- ShapeTest --------------------
int main()
{
    Shape* s1;

    cout << "\n--- Shape ---" << endl;
    s1 = new Shape("Yellow", 200.0);
    s1->display();

    cout << "\n--- Circle ---" << endl;
    s1 = new Circle("Blue", 10.0);
    s1->display();

    cout << "\n--- Rectangle ---" << endl;
    s1 = new Rectangle("Green", 8.0, 6.0);
    s1->display();

    cout << "\n--- Triangle ---" << endl;
    s1 = new Triangle("Purple", 5.0, 7.0);
    s1->display();
}

